### PR TITLE
feat: discard state writes in StateCompute and StateReplay.

### PR DESCRIPTION
**Pull request by raulk**
## Related Issues

This originally started because Glif and other operators begun having blockstore ballooning issues starting with v1.20.3. The issue was the usage of `StateCompute` in the Eth API. That method finishes by flushing the new state to the blockstore. Badger is an LSM + value log store, and it happily accepts every write.

image

## Proposed Changes

The fix is to add a flag to skip writing the state tree when executing a tipset. By default we write, except when the caller requests us not to. Instead of polluting APIs with yet another option, I decided to add an option struct (`ExecutorOpts`). Unfortunately, `TipSetState` is used a many places, so to avoid breakage I introduced a `TipSetWithOpts` method. We _could_ also use the functional options pattern, but those are generally used in more public/user-facing APIs.

This change is no longer necessary for the Eth API, since we no longer need to compute state after https://github.com/filecoin-project/lotus/pull/10446. However, they will be very beneficial to users of `StateCompute` (e.g. exchanges) and `StateReplay`.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

File: chain/stmgr/execute.go [hunk 1]
```diff
@@ -13,6 +13,15 @@ import (
 )
 
 func (sm *StateManager) TipSetState(ctx context.Context, ts *types.TipSet) (st cid.Cid, rec cid.Cid, err error) {
+	return sm.TipSetStateWithOpts(ctx, ts, &TipSetStateOpts{})
+}
+
+type TipSetStateOpts struct {
```
Comment (arajasek) [inline chain/stmgr/execute.go:19]: I'm not convinced this is necessary? With the existing tipsetstate cache, and the newly added fast lookup in https://github.com/filecoin-project/lotus/pull/10445, I think we can drop this. We'll only wind up executing the tipset (making the write) if:

- we're executing a tipset that isn't in our main chain
- we're missing the actual state / receipts root.

**Review by arajasek — COMMENTED** [review]
Broadly looks good to me, and a nice improvement. One note.

**Comment by Stebalien** [comment]
> However, they will be very beneficial to users of StateCompute (e.g. exchanges) and StateReplay.

I'm not so sure. Those users likely _need_ the state.

Is the issue here that we don't check if the blockstore has it before writing?

**Comment by raulk** [comment]
Superseded by #10680, which is simplier and less invasive.

**Comment by eshon** [comment]
FYI - Glif Nodes just reported that archival nodes now seem to be growing at about 19 GiB / day, down from about 50 GiB / day before. (An FEVM archival node is ~3.6 TiB currently.)

**Pull request closed without merging**